From 02a8b628c26f992ebec28cb96d220bccbc6076e1 Mon Sep 17 00:00:00 2001
From: Md Sadre Alam <quic_mdalam@quicinc.com>
Date: Mon, 6 Nov 2023 14:22:54 +0530
Subject: [PATCH 490/500] clk: qcom: ipq9574: Fix crypto bus clock stuck issue

Fix crypto bus clock status stuck issue. In crypto
module removal path clock unprepare will get called
since register address was wrong for crypto bus
clock and crypto ahb clock, due to this its throwing
warning while doing rmmod qcrypto.ko module or while
bootup.

This changes also added .halt_check = BRANCH_HALT_VOTED,
in both bus clock & ahb clock initialization structure.

WARNING:
root@OpenWrt:/# rmmod qcrypto.ko
[   48.121916] ------------[ cut here ]------------
[   48.121940] gcc_crypto_axi_clk status stuck at 'on'
[   48.121964] WARNING: CPU: 0 PID: 5206 at drivers/clk/qcom/clk-branch.c:91 clk_branch_toggle+0x130/0x150

re usbcore nls_base usb_common mii
[   48.405964] CPU: 0 PID: 5206 Comm: rmmod Tainted: G        W          6.1.31 #0
[   48.428193] Hardware name: Qualcomm Technologies, Inc. IPQ9574/RDP418/AP-AL02-C1 (DT)
[   48.435487] pstate: 604000c5 (nZCv daIF +PAN -UAO -TCO -DIT -SSBS BTYPE=--)
[   48.443471] pc : clk_branch_toggle+0x130/0x150
[   48.450240] lr : clk_branch_toggle+0x130/0x150
[   48.454753] sp : ffffff800f3afbd0
[   48.459178] x29: ffffff800f3afbd0 x28: ffffff800f7e6c00 x27: 0000000000000000
[   48.462568] x26: 0000000000000000 x25: 0000000000000000 x24: 0000000000000000
[   48.469687] x23: ffffffc008399ea8 x22: ffffffc00899d72f x21: 0000000000000000
[   48.476805] x20: ffffffc008b5a040 x19: 0000000000000000 x18: 0000000000000000
[   48.483923] x17: 0000000000000000 x16: 0000000000000000 x15: ffffffc008ac9780
[   48.491041] x14: 0000000000000000 x13: 6f27207461206b63 x12: 7574732073757461
[   48.498159] x11: 000000000000023e x10: 00000000ffffffea x9 : 000000000000000c
[   48.505276] x8 : ffffff800f3af958 x7 : 00000000ffffefff x6 : ffffff800f3af960
[   48.512395] x5 : ffffff803ff3b828 x4 : 0000000000000000 x3 : 0000000000000027
[   48.519513] x2 : 0000000000000000 x1 : 0000000000000000 x0 : ffffff800f7e6c00
[   48.526632] Call trace:
[   48.533744]  clk_branch_toggle+0x130/0x150
[   48.536002]  clk_branch2_disable+0x18/0x20
[   48.540170]  clk_core_disable+0x94/0x9c
[   48.544248]  clk_core_disable_lock+0x20/0x34
[   48.547983]  clk_disable+0x20/0x28
[   48.552495]  clk_disable_unprepare+0x14/0x28 [qcrypto]
[   48.555708]  qce_crypto_remove+0x38/0x68 [qcrypto]
[   48.560830]  platform_remove+0x4c/0x64
[   48.565602]  device_remove+0x48/0x6c
[   48.569336]  device_release_driver_internal+0x88/0x140
[   48.573069]  driver_detach+0x80/0x8c
[   48.578016]  bus_remove_driver+0x94/0xbc
[   48.581750]  driver_unregister+0x48/0x4c
[   48.585655]  platform_driver_unregister+0x10/0x18
[   48.589562]  qce_crypto_driver_exit+0x14/0x4c8 [qcrypto]
[   48.594164]  __arm64_sys_delete_module+0x1b4/0x298
[   48.599545]  el0_svc_common.constprop.0+0xc0/0x130
[   48.604146]  do_el0_svc+0x18/0x20
[   48.608918]  el0_svc+0x14/0x3c
[   48.612303]  el0t_64_sync_handler+0xa8/0x134
[   48.615256]  el0t_64_sync+0x14c/0x150
[   48.619684] ---[ end trace 0000000000000000 ]---
[   48.623635] ------------[ cut here ]------------
[   48.627932] gcc_crypto_ahb_clk status stuck at 'on'

Change-Id: Idf8f98c07e73b54e65fc08f146ddd879bee39928
Signed-off-by: Md Sadre Alam <quic_mdalam@quicinc.com>
---
 drivers/clk/qcom/gcc-ipq9574.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/clk/qcom/gcc-ipq9574.c b/drivers/clk/qcom/gcc-ipq9574.c
index 912f4d3c0cdc..68bad53095ae 100644
--- a/drivers/clk/qcom/gcc-ipq9574.c
+++ b/drivers/clk/qcom/gcc-ipq9574.c
@@ -2216,7 +2216,7 @@ static struct clk_rcg2 pcnoc_bfdcd_clk_src = {
 	.halt_reg = 0x16010,
	.halt_check = BRANCH_HALT_VOTED,
 	.clkr = {
-		.enable_reg = 0xb004,
+		.enable_reg = 0xB004,
		.enable_mask = BIT(15),
 		.hw.init = &(const struct clk_init_data) {
 			.name = "gcc_crypto_axi_clk",
@@ -2233,7 +2233,7 @@ static struct clk_branch gcc_crypto_axi_clk = {
 	.halt_reg = 0x16014,
	.halt_check = BRANCH_HALT_VOTED,
 	.clkr = {
-		.enable_reg = 0xb004,
+		.enable_reg = 0xB004,
		.enable_mask = BIT(16),
 		.hw.init = &(const struct clk_init_data) {
 			.name = "gcc_crypto_ahb_clk",
-- 
2.34.1
